SPO-13135 Работа с провайдером

ingvilow opened the pull request:
Пришлось создать новый проект, потому что тот был безвозвратно мной сломан :(

ingvilow committed:
Работа с провайдером

Пришлось создать новый проект, потому что тот был безвозвратно мной сломан :(

arefimenko commented on lib/main.dart:
```diff
@@ -7,109 +11,92 @@ void main() {
-              '$_counter',
+              context.watch<Counter>().counter.toString(),
               style: Theme.of(context).textTheme.headline4,
             ),
```
хоть и вижу что по офф доке `provider` они предлагают делать `watch<Type>()` сразу в месте где это нужно в вёрстке, но по моим ощущениям может ввести в дальнейшем в заблуждение/при рефакторинге будет аффектить смежные вещи/не сразу явно понятно что тут есть подписка на изменения. как вариант решения - `Consumer`, который есть в пакете, благодаря которому ты можешь получить инстанс нотифаера по дженерику и далее работать с ним в билдере при отрисовке

arefimenko commented on lib/provider/counter_provider.dart:
```diff
@@ -0,0 +1,11 @@
+    _count += 1;
+    notifyListeners();
+  }
+}
```
нужно потихоньку нарабатывать привычку писать документацию, даже что тут всё "и так понятно", парой строк всё же стоит покрыть. а чтобы точно не забывать - настроить анализатор [с этим правилом](https://dart-lang.github.io/linter/lints/public_member_api_docs.html), либо использовать пакет `surf_lint_rules`

arefimenko commented on lib/provider/second_page.dart:
```diff
@@ -0,0 +1,54 @@
+            children: [
+              Text(
+                Provider.of<Counter>(context).counter.toString(),
+              ),
```
а точно ли тут происходит работа над новосозданным инстансом счётчика? точно ли он прокинулся в контекст? почему то есть предположения что ты работаешь над первым, который инжектится около MaterialApp

arefimenko replied:
@ingvilow до сих пор не убран лишнее прокидывание зависимости на счётчик

arefimenko reviewed (commented):
по поводу наименования веток/коммитов - как сейчас ведутся задачи в jira, от которых выполняются задания? это отдельные задачи на каждую реализацию или одна общая на всё?
просто вот какие есть договорённости по гитфлоу, кратко опишу основные моменты

наименования веток:
1. формируется из проекта, в котором заведена задача + номер задачи
пример: проект `Surf Studio Practice` (сокращённое скорее всего будет `SSP` (по первым буквам слов)), задача имеет номер `123`
ветка будет начинаться вот так: `SSP-123`, то есть `буквы_проекта`-`номер_таска`
2. краткое описание поставленной задачи
пример: создать экран счётчика - `create_counter_screen`
итог: название ветки получается - `SSP-123_create_counter_screen`
итоговая структура: `буквы-проекта`-`номер-задачи`_`описание_задачи`
где
- буквы проекта - можно взять из задачи
- номер задачи - можно взять из задачи
- описание задачи - краткое описание поставленной задачи, которое полностью отображает основную цель задачи (слова разделяются нижним подчёркиванием, разрешена только латиница, спецсимволы запрещены (могут быть проблемы на разных ОС/гитхабе при работе с ними, исключения - бэкслэш `/`, который можно использовать как группировку веток, которые относятся к одной вещи, чаще это ветки спринтов - `dev/sprint-01` и так далее))

<img width="122" alt="Screenshot 2022-05-13 at 13 09 38" src="https://user-images.githubusercontent.com/54911543/168262352-f47655cd-a310-4591-8539-7480255af346.png">

живой пример со скриншота, ветка будет называться `MFG-33_ci_setup`

наименования коммитов:
1. аналогично названию ветки коммит начинается с описания таска, в нашем случае из текстового примера это `SSP-123`
2. далее коммит содержит в себе описание, согласно [данному гайдлайну](https://www.conventionalcommits.org/en/v1.0.0/)
пример: `SSP-123 feat: counter screen`
заметки: описание коммитов может содержать кириллицу, но в основном это зависит от специфики проекта. один коммит - одно изменение, которое направленно в одну "сторону", то есть относится либо только к вёрстке, либо только к логике и так далее

arefimenko commented:
и ещё краткий комментарий - пр-ы, которые в каком либо виде относятся к ui/ux и работы бизнес логики, которые можно проследить, взаимодействуя с ui - прикрепляем скриншот/видео с отображением работоспособности

arefimenko commented:
ещё не описал пару моментов
1. если задача решена вне рамок какой либо в jira, то вместо `SSP-123` в начало ветки/коммита ставится `no-task`
2. название пр-а в гитхабе должно содежать `SSP-123` + название задачи или `no-task` + краткое описание что было сделано (в идеале чтобы названия веток/пр-ов описывали одну и ту же суть, название пр-а можно на кириллице)

ingvilow committed:
SPO-13135 провайдер

arefimenko commented on lib/main.dart:
```diff
@@ -46,29 +46,33 @@ class _MyHomePageState extends State<MyHomePage> {
-            ),
-          ],
-        ),
+        child: Consumer<Counter>(builder: (_, value, child) {
+          return Column(
+            mainAxisAlignment: MainAxisAlignment.center,
+            children: <Widget>[
+              const Text(
+                'You have pushed the button this many times:',
+              ),
+              Text(
+                value.counter.toString(),
+                style: Theme.of(context).textTheme.headline4,
+              ),
+            ],
+          );
+        }),
```
_минор_ (значит не обязательно править, лишь как информация для размышления)
Благодаря `Consumer` ты можешь точечно обновлять ui. У тебя есть возможность указать `child`, который в свою очередь потом использовать в `builder`, что он не будет перерисовываться на каждый чих билдера. В твоём случае туда можно убрать `Text` виджет

arefimenko reviewed (changes requested):
- приложение не запускается (Flutter 2.10.5, мобильное устройство)
- `FlutterError` при переходе на 2 и 3 экраны с первого экрана
- `AssertionError` при увеличении счётчика с 3 экрана
- поправить подсказки анализатора

ingvilow committed:
SPO-13135 попаравила то, что в пр-е

Только уже не очень помню, что от меня требовалось,но вроде прокинуть через Provider.value как-то

ingvilow committed:
SPO-13135 провайдер

arefimenko commented on lib/main.dart:
```diff
@@ -7,109 +12,101 @@ void main() {
+        }),
+        ChangeNotifierProvider<AnotherCounter>(create: (context) {
+          return AnotherCounter();
+        }),
```
создавать ещё один счётчик - нет необходимости

arefimenko commented on lib/pages/second_page.dart:
```diff
@@ -0,0 +1,53 @@
+class _SecondPageState extends State<SecondPage> {
+  @override
+  Widget build(BuildContext context) {
+    return ChangeNotifierProvider(
+      create: (context) {
+        return Counter();
+      },
```
это по факту тоже лишнее - у тебя создан счётчик у корня приложения. либо он создан для третьего экрана? тогда зачем ещё другой класс счётчика был создан(

arefimenko replied:
@ingvilow не убран лишний провайдер

arefimenko commented on lib/provider_counters/another_counter.dart:
```diff
@@ -0,0 +1,13 @@
+import 'package:flutter/cupertino.dart';
+
+//класс счетчика
+class AnotherCounter extends ChangeNotifier {
```
поведение ничем не отличается от `Counter`, можно было б сделать `extends` поверх его и избавиться от дублирования кода

arefimenko reviewed (commented):
предлагаю созвониться, ибо как вижу есть лёгкое непонимание что происходит

ingvilow committed:
no-task убрала лишнюю реализацю провайдера для счетчика на третьем экране
Поправила ошибки анализатора
